### ISSUE TITLE
Appointment appointment_created_at & appointment_updated_at is not atomic

### ISSUE THREAD

**Issue by francoislepage** [redguava/cliniko-api#232]
Hi there,

I have been using the Cliniko API for workflow integration and it is the second time that when an appointment is created just before the second is about to change, fields updated_at and created_at will not display the same value.

Since the condition updated_at == created_at is the best condition to detect if an object was just initially created and untouched, I believe this is a problem.

Thank you for considering this request.

**Comment by bpinto** [comment]
Could you explain why you need this information?

We can't guarantee that we won't run any asynchronous job after a record is created.

**Comment by francoislepage** [comment]
Hi,

I was using updated_at == created_at to differentiate appointments that just got created and that are still untouched by any further updates.  This is the baseline of the appointment object, that later allow for detecting changes and triggering events.

In my experience, this is true 95% of the times.

Basically, the "lifecycle" of an appointment object goes from "new" when updated_at == created_at, to "updated_field_b" -> "updated_field_c" -> "updated_field_a" and so on, when updated_at != created_at during subsequent pooling.

In my code, **when updated_at != created_at, and I don't have any previous records of this appointment, I consider that I missed something**, so I am "catching-up", hence "unknown status" -> "updated" -> "updated" and so on.  Scenarios that I thought would lead to this outcome are downtimes and  initial software deployment with no previous records.

In other words, when updated_at is never equal to created_at at some point in time, I do not feel like I have a solid baseline to follow-up with changes on a specific appointment.

I could fix it in the client code by considering that if the two values are only a few seconds appart, I make the appointment look brand new, but I feel like this would be a "patch".

According to me, the proper behaviour would be to make sure these values are equal when the appointment is initially created.

I hope this clarifies.

Thank you :-)

François

**Comment by bpinto** [comment]
> According to me, the proper behaviour would be to make sure these values are equal when the appointment is initially created.

Let's have image upload as an example, when you upload an image (not available through the API yet), we need to process the uploaded image so that it's stored in different sizes and we will use the proper one depending on the device you are using. This process is done asynchronously, and we will update the record once these asynchronous jobs finish.

---

> In my code, when updated_at != created_at, and I don't have any previous records of this appointment, I consider that I missed something, so I am "catching-up", hence "unknown status" -> "updated" -> "updated" and so on. Scenarios that I thought would lead to this outcome are downtimes and initial software deployment with no previous records.

I'm not sure why you are recording every change to a record, instead of just using what we return from the API. Let's say a user updates the appointment twice, in less than 5 seconds, you are going to miss the first change anyway.

**Comment by francoislepage** [comment]
Hi,

Thanks for your answer. I do understand your point about asynchronous updates e.g a "reminder sent" flag added to the record for instance, right after the initial creation.

You are also totally correct about the fact that if a user [receptionist] changes a record in a time frame smaller than my polling interval, created_at won't be equal than updated_at on the very first poll.

That's why I have asked my admin staff to delete appointments and re-create them if a change is required less than 15 min after an initial appointment is being booked.  This is *my* ugly patch.  I eventually need to fix my code so changes within the first polling interval are tolerated.

Although you correctly highlighted flaws to my approach, the point I'm trying to make is that in the initial transaction (the initial appointment "save" before anything happens to it) created_at should equal updated_at.  Architecturally speaking, I truly believe this is what makes the most sense.  The "one second differences" I noticed now 3 times in the last week are definitely not due to an asynchronous process.

Maybe you are resolving in the code the current date and time "two times" within the same "transaction"? e.g. calling twice LocalDateTime.now() rather than storing the first returned value in a variable???  Just an hypothesis on why this behaviour could happen in such a small amount of occurrences.

Thank you

François

**Comment by bpinto** [comment]
> Maybe you are resolving in the code the current date and time "two times" within the same "transaction"? e.g. calling twice LocalDateTime.now() rather than storing the first returned value in a variable??? Just an hypothesis on why this behaviour could happen in such a small amount of occurrences.

The reason why the appointment `created_at` and `updated_at` doesn't always match has to do with a cached column we have on our appointments table that caches the `patient_ids` for the appointment record. While it would be possible for us to optimize it for individual appointments, the same is not true for group appointments, where adding new patients to an appointment would cause the `updated_at` on the group appointment to bump.

While that optimization is probably something that we are going to do eventually, I would like to help you with your code to not rely on this.

P.S.: Are you aware of the "View log" button on patients -> appointments?

**Comment by francoislepage** [comment]
Hi,

While I did not really understand your explanation of the cached column and the link with patient_ids, what I am getting from it is that there are some structural reasons other than the very simple hypothesis I provided that explains why I can't trust create_at == updated_at. :-)

Thanks for agreeing on optimizing the individual appointment eventually.  On my end, I will follow your advise and perform the update to my code so that I am not dependant anymore on this condition.

I had never used the View Log.  That's a good one, thank you.

François
